File: 45.cpp
//
// Created by DuLi on 2023/10/17.
//
#include <iostream>
using std::vector;

class Solution {
public:
    int jump(vector<int>& nums) {
        int farthest = 0;
        int end = 0;
        int steps = 0;

        for (int i = 0; i < nums.size() - 1; ++i) {
            farthest = farthest > nums[i] + i ? farthest : nums[i] + i;
            if (i == end){
                steps ++;
                end = farthest;
            }
        }
        return steps;
    }
};

File: 45.py
def jump(nums):
    if not nums or len(nums) == 1:
        return 0

    jumps, end, farthest = 0, 0, 0
    for i in range(len(nums) - 1):
        farthest = max(farthest, i + nums[i])
        if i == end:
            jumps += 1
            end = farthest

    return jumps

nums = [2,3,1,1,4]
print(jump(nums))